[MAPS3D-425] introduce skirts for globe rendering

alexey-romanov opened the pull request:
Absent skirts lead to holes for hight exaggerated terrain in globe mode.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [x] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>introduce skirts for terrain globe mode</changelog>`

alexey-romanov commented:
Introduce strips for terrain globe rendering. Absent strips lead to holes for high exaggerations.

Before:
![before_no_strips](https://user-images.githubusercontent.com/1542101/212902142-1664cb97-69b4-4517-a48e-9ba769b3ec1c.png)
After:
![after_with_strips](https://user-images.githubusercontent.com/1542101/212902160-ae93e6ff-123a-4e7e-be94-0676d6a974e7.png)

Before:
<img width="1512" alt="before_holes" src="https://user-images.githubusercontent.com/1542101/212902311-4c3fd8bc-a0f0-4c29-8b5a-e6c8db3e7fb2.png">

After:

<img width="1487" alt="after_no_holes" src="https://user-images.githubusercontent.com/1542101/212902339-47d3d1a0-bc1e-426c-a6e0-42dcb1394415.png">

alexey-romanov commented:
Case with transparent textures not handled -> update with moving skirts to be rendered last will be made

mourner commented on src/geo/projection/globe_util.js:
```diff
@@ -724,37 +730,117 @@ export class GlobeSharedBuffers {
+    //     gridIndices: TriangleIndexArray,
+    //     gridSegments: Array<SegmentVector>
+    // };
+    _fillGridMeshWithLods(longitudinalCellsCount: number, latitudinalLods: number[], hasStrip: boolean): GridWithLods {
```
Nit: should we eliminate the `hasStrip` flag since it's always true? (I guess this a debugging leftover?)

alexey-romanov replied:
I have removed the flag from the interface, but kept it internal for potential debug uses

mourner commented on src/geo/projection/globe_util.js:
```diff
@@ -724,37 +730,117 @@ export class GlobeSharedBuffers {
+        for (let y = 0; y < yVerticesHighLodNoStrip; ++y) {
+            for (let x = 0; x < xVertices; ++x) {
+                const isSideBorder = (x === 0 || x === xVertices - 1);
+                let stripFactor = 0;
+                if (isSideBorder && hasStrip) {
+                    stripFactor = 1;
                 }
+
+                vertices.emplaceBack(xCoordAdjusted(x), y, stripFactor);
```
Nit: we could use some ternaries in a few places here to make the code simpler, e.g.:

```suggestion
                vertices.emplaceBack(xCoordAdjusted(x), y, isSideBorder && hasStrip ? 1 : 0);
```

alexey-romanov replied:
done

mpulkki-mapbox commented on src/terrain/draw_terrain_raster.js:
```diff
@@ -171,6 +171,10 @@ function drawTerrainForGlobe(painter: Painter, terrain: Terrain, sourceCache: So
     batches.forEach(isWireframe => {
+        const tr = painter.transform;
+        // Use varying skirt hight applying limit for extremely large axaggerations
+        const skirtHeightValue = Math.min(skirtHeight(tr.zoom) * terrain.exaggeration(), 20000);
+
```
How was the value "20000" chosen?

alexey-romanov replied:
heuristically, without the limiting factor on certain zoom levels, e.g. 1.5 +- there can be seen triangles penetrating the earth surface. It was revealed by the globe anti-aliasing test with extreme exaggerations

mpulkki-mapbox replied:
on the opposite side of the globe? What are the implications of triangles penetrating the surface?

alexey-romanov replied:
Fixed

mpulkki-mapbox commented on src/terrain/draw_terrain_raster.js:
```diff
@@ -171,6 +171,10 @@ function drawTerrainForGlobe(painter: Painter, terrain: Terrain, sourceCache: So
 
     batches.forEach(isWireframe => {
+        const tr = painter.transform;
+        // Use varying skirt hight applying limit for extremely large axaggerations
```
"hight" typo?

mpulkki-mapbox commented on src/geo/projection/globe_util.js:
```diff
@@ -724,37 +731,117 @@ export class GlobeSharedBuffers {
+                    indices.emplaceBack(idx + offsetToNextRow, idx + offsetToNextRow + 1, idx + 1);
+                }
+            }
+            segments.push(SegmentVector.simpleSegment(0, indexOffset, vertices.length, indices.length - indexOffset));
```
Are you including skirt triangles into the same segment as the main grid triangles? We could have two different segments: one with skirts and one without skirts. This way we can skip rendering of skirts if the terrain is not enabled in globe view.

mpulkki-mapbox replied:
If I understood correctly this change will always render skirts?

alexey-romanov replied:
yes, always

alexey-romanov replied:
Otherwise this would mean sending more draw routines to the graphics pipeline. I think it is either

same count of render calls even if that means rendering +6% extra degenerate triangles

or 

double the amount of render calls

mpulkki-mapbox replied:
would it be possible to construct index buffer so that skirt triangles are immediately after terrain triangles? This wouldn't increase draw count as the only difference would be the number of rendered indices?

alexey-romanov replied:
currently, always, but yeah two segments (one with skirt and one without) can be used depending on terrain exaggeration. I'll do that modification

alexey-romanov replied:
Done

mpulkki-mapbox commented on src/shaders/_prelude_terrain.vertex.glsl:
```diff
@@ -24,6 +24,18 @@ vec3 elevationVector(vec2 pos) { return vec3(0, 0, 1); }
 
+// Handle skirt flag for terrain & globe shaders
+
+const float skirtOffset = 24575.0;
+vec3 decomposeToPosAndSkirt(vec2 posWithComposedSkirt)
+{
+    float skirt = float(posWithComposedSkirt.x >= skirtOffset);
+    vec2 pos = posWithComposedSkirt - vec2(skirt * skirtOffset, 0.0);
+
+    return vec3(pos, skirt);
+}
```
Nice, this increases readability quite a lot! :+1:

mpulkki-mapbox commented on src/terrain/draw_terrain_raster.js:
```diff
@@ -254,7 +258,7 @@ function drawTerrainForGlobe(painter: Painter, terrain: Terrain, sourceCache: So
                     globeRasterUniformValues(tr.projMatrix, poleMatrix, poleMatrix, normalizeMatrix, 0.0, mercatorCenter,
                     tr.frustumCorners.TL, tr.frustumCorners.TR, tr.frustumCorners.BR, tr.frustumCorners.BL,
-                    tr.globeCenterInViewSpace, tr.globeRadius, viewport), "globe_pole_raster", vertexBuffer,
+                    tr.globeCenterInViewSpace, tr.globeRadius, viewport, 0), "globe_pole_raster", vertexBuffer,
```
Skirt is not required for pole geometry?

mpulkki-mapbox replied:
As discussed offline, there's no LOD change between pole triangle fans and their neighbouring proxy tiles, hence no skirt required.

mpulkki-mapbox reviewed (commented):
I noticed some areas where the skirt might be missing.

https://user-images.githubusercontent.com/55925868/213457457-d88c7c99-6c9b-4095-b0f8-841f5c5b2d2a.mov

mpulkki-mapbox commented:
Thanks for addressing the review comments @alexey-romanov! Code changes looks good to me but unfortunately there are still some scenarios where the skirt geometry is missing. It is also possible to see gaps when zooming in and out rapidly as I wrote previously.

<img width="460" alt="image" src="https://user-images.githubusercontent.com/55925868/213710872-87b05842-fc4e-4385-a9f7-1d118539c91e.png">

I would suggest fixing remaining issues where gaps are visible even when camera is not moving. We can defer rest of the cases to a follow-up PR.

alexey-romanov commented:
>

alexey-romanov commented:
> Thanks for addressing the review comments @alexey-romanov! Code changes looks good to me but unfortunately there are still some scenarios where the skirt geometry is missing. It is also possible to see gaps when zooming in and out rapidly as I wrote previously.
> 
> <img alt="image" width="460" src="https://user-images.githubusercontent.com/55925868/213710872-87b05842-fc4e-4385-a9f7-1d118539c91e.png">
> 
> I would suggest fixing remaining issues where gaps are visible even when camera is not moving. We can defer rest of the cases to a follow-up PR.

I was able to repro it locally. It is related to the case of transition between globe and mercator

alexey-romanov commented:
> Thanks for addressing the review comments @alexey-romanov! Code changes looks good to me but unfortunately there are still some scenarios where the skirt geometry is missing. It is also possible to see gaps when zooming in and out rapidly as I wrote previously.
> 
> <img alt="image" width="460" src="https://user-images.githubusercontent.com/55925868/213710872-87b05842-fc4e-4385-a9f7-1d118539c91e.png">
> 
> I would suggest fixing remaining issues where gaps are visible even when camera is not moving. We can defer rest of the cases to a follow-up PR.

Should be fixed now
